File: leetcode/518.cpp
// https://leetcode.com/problems/coin-change-ii/
#include <iostream>
#include <cstring>
#include <stdlib.h>
#include <vector>
#include <unordered_map>
#include <math.h>

using std::vector;
using std::cout;
using std::cin;
using std::endl;
using std::string;
using std::max;

class Solution {

public:
    int change(int amount, vector<int>& coins) {
        vector<int> seen(amount+1,0);
        seen[0] = 1;
        for (auto &&i : coins)
        {
            for (int j = 1; j < amount+1; j++)
            {
                if(j-i >= 0)seen[j]+= seen[j-i];
            }
            
        }
        return seen[amount];
        
    }
};

int main(){
    Solution S;
    vector<int> coins = {1,2,5};
    cout << S.change(5,coins) << endl;
    return 0;
}

File: leetcode/518.py
# https://leetcode.com/problems/coin-change-ii/
from typing import *
class Solution:
    def change(self, amount: int, coins: List[int]) -> int:
        dp = [0] * (amount + 1)
        dp[0] = 1
        coins.sort(reverse=True)
        for i in coins:
            for j in range(1, amount+1):
                if j-i>=0:
                    dp[j] += dp[j-i]
        return dp[-1]



S = Solution()
X = S.change(amount = 5, coins = [1,2,5])
print(X)